// nix-clap/dist-esm/clap-node.js
/**
 * Represents a node in a command-line application parser.
 *
 * @remarks
 * This class is used to represent commands or options in a CLI application. Each node can have a parent node,
 * which represents a hierarchical relationship (e.g., command -> sub-command, option).
 *
 * @example
 * ```typescript
 * const rootNode = new ClapNode('root', 'rootAlias');
 * const childNode = new ClapNode('child', 'childAlias', rootNode);
 * rootNode.addArg('arg1');
 * childNode.addError(new Error('An error occurred'));
 * ```
 *
 * @public
 */
import { _PARENT } from "./symbols.js";
export class ClapNode {
    /**
     *
     * @param name - idiomatic name of the command
     * @param inputName - name that was specified by user to trigger this command
     * @param parent
     */
    constructor(name, alias, parent) {
        this.name = name;
        this.alias = alias;
        this.argsList = [];
        this.argv = [];
        this.argsMap = {};
        if (parent) {
            Object.defineProperty(this, _PARENT, {
                value: parent,
                configurable: true,
                enumerable: false,
                writable: true
            });
        }
        this._errors = [];
    }
    /**
     * Add an arg to this node
     * @param arg
     * @returns
     */
    addArg(arg) {
        this.argsList.push(arg);
    }
    addVerbatimArg(arg) {
        this.argv.push(arg);
    }
    addError(e) {
        this._errors.push(e);
    }
    get hasErrors() {
        return this._errors.length > 0;
    }
    get error() {
        return this._errors[0];
    }
    get errors() {
        return this._errors;
    }
    get parent() {
        return this[_PARENT];
    }
    getParent() {
        return this.parent;
    }
}
