// nix-clap/dist-cjs/base.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliBase = exports.UnknownCliArgError = exports.UnknownOptionError = exports.InvalidArgSpecifierError = exports.isRootCommand = exports.rootCommandName = void 0;
const assert_1 = __importDefault(require("assert"));
const xtil_ts_1 = require("./xtil.cjs");
const SUPPORT_TYPES = ["number", "string", "float", "boolean", "int", "integer"];
exports.rootCommandName = "~root-command~";
const isRootCommand = (name) => name === exports.rootCommandName;
exports.isRootCommand = isRootCommand;
/**
 * Error thrown when an invalid argument specifier is encountered.
 *
 * @extends {Error}
 */
class InvalidArgSpecifierError extends Error {
    constructor(msg, arg = "") {
        super(msg);
        this.arg = arg;
    }
}
exports.InvalidArgSpecifierError = InvalidArgSpecifierError;
/**
 * User provided an option that's unknown
 */
class UnknownOptionError extends Error {
    constructor(msg, data) {
        super(msg);
        this.data = data;
    }
}
exports.UnknownOptionError = UnknownOptionError;
/**
 * User provided an cli argument that's unknown
 */
class UnknownCliArgError extends Error {
    constructor(msg, arg) {
        super(msg);
        this.arg = arg;
    }
}
exports.UnknownCliArgError = UnknownCliArgError;
/**
 * The base class for objects to represent the processed specification of a
 * CLI command or option.
 */
class CliBase {
    /**
     * Creates a new CliBase instance.
     * @param name - The name of the command or option
     * @param spec - The specification for the command or option
     */
    constructor(name, spec) {
        this.name = name;
        this.args = [];
        this.spec = spec;
        this.needArgs = 0;
        this.expectArgs = 0;
    }
    /** Indicates whether this command or option has a single argument */
    get isSingleArg() {
        return this.args.length === 1;
    }
    /** Indicates whether this command or option has any arguments */
    get hasArgs() {
        return this.args.length > 0;
    }
    /** Returns the CLI type (e.g., "command" or "option") */
    get cliType() {
        return "command";
    }
    /** Indicates whether this command or option has variadic arguments */
    get isVariadicArgs() {
        return !!this.variadic;
    }
    /**
     * Processes the argument specification string and populates the args array.
     * @throws {InvalidArgSpecifierError} If the argument specification is invalid
     */
    processArgs() {
        const spec = this.spec;
        const args = spec.args;
        if (!args) {
            return;
        }
        let seenOptional;
        //
        // check "<arg1> <arg2> [arg3]"
        //
        args.replace(/([<[])([^>\]]*)([>\]])/g, (a, mark, xname) => {
            var _a;
            (0, assert_1.default)(!this.variadic, `For args specifier of '${this.name}', only the last one can be variadic`);
            // "[string foo..1,3]"
            //   xname = xname.trim();
            const xm = xname.match(/([^\. ]+)?( +)?([^\.]+)?(\.\.+)?([^\.,]*)?( *, *)?([^\.]+)?$/);
            (0, assert_1.default)(xm, new InvalidArgSpecifierError(`Invalid args specifier '${xname}' for '${this.name}'.`));
            let variadic;
            let min = 1;
            let max = 1;
            const name = xm[1]; // "a b..1,3" <-- "a"
            const type = xm[3] || "string"; // "a b..1,3" <-- "b"
            (0, assert_1.default)(SUPPORT_TYPES.indexOf(type) >= 0 ||
                ((_a = (spec.customTypes ||
                    // @ts-expect-error - coercions is deprecated legacy property name, kept for backward compatibility
                    spec.coercions)) === null || _a === void 0 ? void 0 : _a.hasOwnProperty(type)), new InvalidArgSpecifierError(`${this.cliType} ${this.name} - Unknown type '${type}' for argument '${a}'.`, args));
            if (xm[5]) {
                // "a b..1,3" : "1"
                min = max = (0, xtil_ts_1.validParseInt)(xm[5], 0);
                if (xm[7]) {
                    // "a b..1,3" : "3"
                    max = (0, xtil_ts_1.validParseInt)(xm[7], min);
                }
                else if (xm[6]) {
                    // "a b..1,3" : ","
                    max = Infinity;
                }
            }
            else if (xm[4]) {
                // "a b..1,3" : ".."
                min = 0;
                max = Infinity;
            }
            const required = mark === "<";
            if (required) {
                (0, assert_1.default)(!seenOptional, new InvalidArgSpecifierError(`Required arg '${xname}' cannot follow optional arg '${seenOptional}'.`));
                this.needArgs += min;
            }
            else {
                seenOptional = xname;
            }
            if (max > min) {
                variadic = true;
                this.variadic = { min, max };
                this.expectArgs += max;
            }
            else {
                this.expectArgs += min;
            }
            (0, assert_1.default)(!SUPPORT_TYPES.includes(name), new InvalidArgSpecifierError(`For args of ${this.name} - Its name is using a type name: ${name}.`));
            this.args.push({
                required,
                name,
                type,
                variadic,
                min,
                max
            });
            return "";
        });
        (0, assert_1.default)(!(args && this.args.length === 0), new InvalidArgSpecifierError(`Invalid args specifier '${args}' for '${this.name}'.`));
    }
}
exports.CliBase = CliBase;
