// nix-clap/dist-cjs/command-base.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unknownCommandBaseNoOptions = exports.unknownCommandBase = exports.CommandBase = void 0;
const base_ts_1 = require("./base.cjs");
const options_ts_1 = require("./options.cjs");
const xtil_ts_1 = require("./xtil.cjs");
/**
 * Represents a command in the CLI application.
 *
 * @template CommandSpec - The specification type for the command.
 *
 * @extends CliBase<CommandSpec>
 *
 * @property {Record<string, string>} subAliases - A record of sub-command aliases.
 * @property {Options} options - The options associated with the command.
 * @property {number} subCommandCount - The count of sub-commands.
 * @property {Record<string, Command>} subCommands - A record of sub-commands.
 * @property {number} execCount - The count of executable sub-commands.
 * @property {Command} [parent] - The parent command, if any.
 *
 * @remarks
 * This class provides methods to manage and execute CLI commands, including handling sub-commands, options,
 * and aliases. It also includes functionality to generate help messages and verify the uniqueness of options
 * and aliases.
 *
 * @example
 * ```typescript
 * const cmdSpec: CommandSpec = { /* command specification *\/ };
 * const command = new Command('start', cmdSpec);
 * command.setCommandAliases(['run', 'execute'], 'start');
 * const helpMessage = command.makeHelp('myProgram');
 * console.log(helpMessage.join('\n'));
 * ```
 */
class CommandBase extends base_ts_1.CliBase {
    /**
     * Creates an instance of a Command.
     *
     * @param name - The name of the command.
     * @param cmdSpec - The specification object for the command.
     * @param ncConfig - The NixClap configuration.
     * @param parent - An optional parent command.
     */
    constructor(name, cmdSpec, ncConfig, parent) {
        const specCopy = (0, xtil_ts_1.dup)(cmdSpec);
        super(name, specCopy);
        this.subCmdCount = 0;
        this.execCount = 0;
        this.subCmdsBase = {};
        this.subAliases = {};
        this.parent = parent;
        this.ncConfig = ncConfig;
        specCopy.alias = [].concat(specCopy.alias || []);
        specCopy.desc = specCopy.desc;
        this.options = new options_ts_1.Options(specCopy.options, this);
        this.needArgs = 0;
        this.expectArgs = 0;
        this.processArgs();
        // let count = 0;
        if (specCopy.subCommands) {
            for (const subName in specCopy.subCommands) {
                const subSpec = specCopy.subCommands[subName];
                this.subCmdCount++;
                this.subCmdsBase[subName] = new CommandBase(subName, subSpec, this.ncConfig, this);
                // count++;
                if (subSpec.exec) {
                    this.execCount++;
                }
            }
        }
        if (parent) {
            this.verifyNoDupOptions();
            parent.setCommandAliases(specCopy.alias, name);
        }
    }
    /**
     * Sets aliases for a command.
     *
     * @param alias - An array of alias strings to be associated with the command.
     * @param name - The name of the command to which the aliases will be assigned.
     * @throws {Error} Throws an error if an alias is already used by another command.
     */
    setCommandAliases(alias, name) {
        alias.forEach(a => {
            if (this.subAliases[a]) {
                throw new Error(`Command ${name} alias ${a} already used by command ${this.subAliases[a]}`);
            }
            this.subAliases[a] = name;
        });
    }
    /**
     * Verifies that an option with the given name does not already exist in the current command or any of its
     * parent commands. If the option exists, an error is thrown.
     *
     * @param optName - The name of the option to verify.
     * @param _from - The command from which the verification is initiated.
     * @throws {Error} If the option already exists in the current command or any parent command.
     */
    verifyOptionNotExist(optName, _from) {
        // Skip check for 'help' option - it's intentionally inherited by all subcommands
        if (optName === "help")
            return;
        if (_from !== this && this.options._options.hasOwnProperty(optName)) {
            throw new Error(`Command ${_from.name} option ${optName} already used by parent command '${this.name}'`);
        }
        if (this.parent) {
            this.parent.verifyOptionNotExist(optName, _from);
        }
    }
    /**
     * Verifies that there are no duplicate options in the command.
     * Iterates through the options and checks each one to ensure it does not already exist.
     * Throws an error if a duplicate option is found.
     */
    verifyNoDupOptions() {
        for (const _key in this.options._options) {
            this.verifyOptionNotExist(_key, this);
        }
    }
    /**
     * Get the allowUnknownOption flag
     *
     * @returns {boolean | undefined} The allowUnknownOption flag
     */
    get allowUnknownOption() {
        return this.cmdSpec.allowUnknownOption;
    }
    /**
     * Get the verbatim arguments
     *
     * @returns {string} The verbatim arguments
     */
    get verbatimArgs() {
        return this.cmdSpec.args || "";
    }
    get desc() {
        return (0, xtil_ts_1.cbOrVal)(this.cmdSpec.desc);
    }
    get usage() {
        return (0, xtil_ts_1.cbOrVal)(this.cmdSpec.usage);
    }
    get exec() {
        return this.cmdSpec.exec;
    }
    get alias() {
        return [].concat(this.cmdSpec.alias);
    }
    /**
     * Retrieves the command specification.
     *
     * @returns {CommandSpec} The specification of the command.
     */
    get cmdSpec() {
        return this.spec;
    }
    /**
     * Matches a sub-command based on the provided alias.
     *
     * @param alias - The alias of the sub-command to match.
     * @returns An object containing the matched command's name, alias, and the command itself.
     *
     * @remarks
     * If the alias does not directly match a sub-command, it will attempt to match using sub-aliases.
     * If no match is found, the alias is returned as the name with an undefined command.
     *
     * @example
     * ```typescript
     * const result = matchSubCommand('start');
     * console.log(result.name); // 'start'
     * console.log(result.alias); // 'start'
     * console.log(result.cmd); // Command object or undefined
     * ```
     */
    matchSubCommand(alias) {
        let cmd = this.subCmdsBase[alias];
        let name = alias;
        if (!cmd) {
            name = this.subAliases[alias];
            if (name) {
                cmd = this.subCmdsBase[name];
            }
            else {
                name = alias;
            }
        }
        return {
            name,
            alias: alias,
            cmd
        };
    }
    /**
     * Get the number of commands, including sub commands, that has exec
     *
     * @returns
     */
    getExecCount() {
        let a = this.exec ? 1 : 0;
        for (const cmd in this.subCmdsBase) {
            a += this.subCmdsBase[cmd].getExecCount();
        }
        return a;
    }
    /**
     * Generates a help message for the command and its subcommands.
     *
     * @param progName - Optional name of the program to be included in the help message.
     * @returns An array of strings representing the formatted help message.
     */
    makeHelp(progName) {
        if (progName) {
            progName = `${progName} `;
        }
        else if (!(0, base_ts_1.isRootCommand)(this.alias[0])) {
            progName = `${this.name} `;
        }
        else {
            // we want to skip showing the root command name
            progName = "";
        }
        const data = [];
        for (const [name, cmdBase] of Object.entries(this.subCmdsBase)) {
            let args = cmdBase.verbatimArgs;
            args = args ? ` ${args}` : "";
            const strs = [`${progName}${name}${args}`, cmdBase.desc ? ` ${cmdBase.desc.trim()}` : ""];
            const alias = cmdBase.alias.join(" ");
            strs.push(alias.length > 0 ? `[aliases: ${alias}]` : "");
            data.push(strs);
        }
        const cmdWidth = data.reduce((max, n) => (n[0].length > max ? n[0].length : max), 0);
        (0, xtil_ts_1.resetZebraIndex)();
        return data.reduce((help, strs) => {
            const lines = (0, xtil_ts_1.fitLines)(strs, "  ", "    ", cmdWidth, (0, xtil_ts_1.getTerminalWidth)());
            (0, xtil_ts_1.nextZebraIndex)();
            return help.concat(lines);
        }, []);
    }
}
exports.CommandBase = CommandBase;
/**
 * Unknown command base with options.  This is used when allowUnknownOption is true.
 */
exports.unknownCommandBase = new CommandBase("unknown", {
    options: {},
    subCommands: {},
    allowUnknownOption: true,
    desc: "Unknown command",
    usage: "Unknown command",
    alias: []
});
/**
 * Unknown command base with no options.  This is used when allowUnknownOption is false.
 */
exports.unknownCommandBaseNoOptions = new CommandBase("unknown no unknown options", {
    options: {},
    subCommands: {},
    allowUnknownOption: false,
    desc: "Unknown command",
    usage: "Unknown command",
    alias: []
});
