// nix-clap/dist-cjs/options.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Options = void 0;
const assert_1 = __importDefault(require("assert"));
const option_base_ts_1 = require("./option-base.cjs");
const xtil_ts_1 = require("./xtil.cjs");
/**
 * The `Options` class is responsible for managing and processing command-line options.
 * It handles the specification data for options, including creating instances of options,
 * managing aliases, and generating help text for the options.
 *
 * @remarks
 * This class is designed to work with a command-line interface, providing functionality
 * to match provided option data against available options and aliases, and to generate
 * formatted help text for the options.
 *
 * @example
 * ```typescript
 * const options = new Options(specGroup, command);
 * const matchedOption = options.match(optionData);
 * const helpText = options.makeHelp();
 * ```
 *
 * @public
 */
class Options {
    constructor(specGroup, cmd) {
        this.specData = (0, xtil_ts_1.dup)(specGroup);
        this._optAlias = {};
        this._options = {};
        this.command = cmd;
        this.processSpecData();
    }
    /**
     * Processes the specification data for options.
     *
     * This method iterates over the `specData` object, creating new `Option` instances
     * for each entry and storing them in the `_options` property. It also handles
     * aliases for options, ensuring that each alias is unique and not already used
     * by another option.
     *
     * @throws {Error} If an alias is already used by another option.
     */
    processSpecData() {
        this.count = 0;
        for (const [name, _spec] of Object.entries(this.specData)) {
            this.count++;
            this._options[name] = new option_base_ts_1.OptionBase(name, _spec);
            if (_spec.alias) {
                if (!Array.isArray(_spec.alias)) {
                    _spec.alias = [_spec.alias];
                }
                _spec.alias.forEach(a => {
                    (0, assert_1.default)(!this._optAlias.hasOwnProperty(a), `Init command ${this.command.name} failed - Option alias ${a} already used by option ${this._optAlias[a]}`);
                    this._optAlias[a] = name;
                });
            }
        }
    }
    /**
     * Matches the provided option data against the available options and aliases.
     *
     * @param data - The option data to match.
     * @returns The matched option data with updated name and alias, or `false` if no match is found.
     */
    match(data) {
        const alias = data.name;
        let name;
        let option = this._options[alias];
        if (option) {
            name = alias;
        }
        else if (this._optAlias[alias]) {
            name = this._optAlias[alias];
            option = this._options[name];
        }
        else {
            return false;
        }
        return { ...data, name, alias: alias, option };
    }
    /**
     * Finds the length of the longest help text among all options.
     *
     * @returns {number} The length of the longest help text.
     */
    findLongestOptionHelpText() {
        let max = 0;
        for (const opt of Object.values(this._options)) {
            max = Math.max(opt.help.length, max);
        }
        return max;
    }
    /**
     * Generates a help text for the available options.
     *
     * This method constructs a formatted help text by iterating over the available options,
     * appending their descriptions, types, and default values if applicable. The help text
     * is formatted to fit within a specified width. Alternating options are dimmed for
     * better readability.
     *
     * @returns {string[]} An array of strings representing the formatted help text.
     */
    makeHelp() {
        const width = this.findLongestOptionHelpText();
        let help = [];
        (0, xtil_ts_1.resetZebraIndex)();
        for (const opt of Object.values(this._options)) {
            const tail = [];
            if (opt.type) {
                tail.push(`[${opt.type}]`);
            }
            if (opt.spec.argDefault && opt.isSingleArg) {
                tail.push(`[default: ${JSON.stringify(opt.spec.argDefault)}]`);
            }
            const desc = ((0, xtil_ts_1.cbOrVal)(opt.spec.desc) || "").trim();
            const strs = [opt.help, desc ? ` ${desc}` : "", tail.filter(x => x).join(" ")];
            const lines = (0, xtil_ts_1.fitLines)(strs, "  ", "    ", width, (0, xtil_ts_1.getTerminalWidth)());
            help = help.concat(lines);
            (0, xtil_ts_1.nextZebraIndex)();
        }
        return help;
    }
}
exports.Options = Options;
