// nix-clap/dist-cjs/option-base.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optUnknown = exports.OptionBase = exports.OPTION_FIELDS = exports.SUPPORT_TYPES = void 0;
const base_ts_1 = require("./base.cjs");
const xtil_ts_1 = require("./xtil.cjs");
exports.SUPPORT_TYPES = ["count", "string", "number", "float", "boolean"];
exports.OPTION_FIELDS = {
    alias: ["string", "array"],
    type: ["string"],
    desc: ["string", "function"],
    describe: ["string", "function"],
    description: ["string", "function"],
    default: [],
    require: ["boolean"],
    requireArg: ["boolean"],
    allowCmd: ["array"],
    customTypes: ["object"]
};
/**
 *
 */
// export type OptionData = {
//   name: string;
//   value: string;
//   verbatim: string;
//   dashes: number;
// };
/**
 * The class to represent an CLI Option from the specification of an option
 */
class OptionBase extends base_ts_1.CliBase {
    constructor(name, optSpec) {
        const specCopy = (0, xtil_ts_1.dup)(optSpec);
        super(name, specCopy);
        this.processArgs();
        this.type = this._buildTypeString();
        this.help = this._buildHelpString(name, optSpec);
    }
    /**
     * Builds the type string for display in help text.
     * Returns empty string for no args, single type for one arg, or "type .." for variadic.
     */
    _buildTypeString() {
        if (this.expectArgs === 0) {
            return "";
        }
        else if (this.expectArgs === 1) {
            return this.args[0].type;
        }
        else {
            return `${this.args[0].type} ..`;
        }
    }
    /**
     * Builds the help string showing option name and aliases.
     * Format: "--name, -a, --alias"
     */
    _buildHelpString(name, optSpec) {
        return [`--${name}`]
            .concat([]
            .concat(optSpec.alias)
            .filter(x => x)
            .map(xtil_ts_1.prefixOption))
            .join(", ");
    }
    get isCounting() {
        return this.spec.counting !== undefined;
    }
    get cliType() {
        return "option";
    }
    get unknown() {
        return false;
    }
    get isUnknown() {
        return false;
    }
}
exports.OptionBase = OptionBase;
class UnknownOption extends OptionBase {
    get unknown() {
        return true;
    }
    get isUnknown() {
        return true;
    }
}
exports.optUnknown = new UnknownOption("", {});
